x/evm/keeper/state_transition.go

```diff
@@ -321,30 +342,49 @@ func (k *Keeper) ApplyMessageWithConfig(
 			return nil, errorsmod.Wrap(err, "failed to apply state override")
 		}
 	}
-	evm = k.NewEVM(ctx, msg, cfg, stateDB)
+	tracingStateDB := vm.StateDB(stateDB)
+	if hooks := cfg.Tracer; hooks != nil {
+		tracingStateDB = statedb.NewHookedState(stateDB, hooks)
+	}
+	evm = k.NewEVM(ctx, msg, cfg, tracingStateDB)
 	// Allow the tracer captures the tx level events, mainly the gas consumption.
 	leftoverGas := msg.GasLimit
-	sender := vm.AccountRef(msg.From)
+	sender := msg.From
 	tracer := cfg.GetTracer()
 	debugFn := func() {
 		if tracer != nil && cfg.DebugTrace {
-			stateDB.AddBalance(sender.Address(), new(big.Int).Mul(msg.GasPrice, new(big.Int).SetUint64(leftoverGas)))
+			stateDB.AddBalance(sender, uint256.NewInt(0).Mul(uint256.MustFromBig(msg.GasPrice), uint256.NewInt(leftoverGas)), tracing.BalanceChangeUnspecified)
 		}
 	}
 	if tracer != nil {
+		if tracer.OnGasChange != nil {
+			tracer.OnGasChange(0, msg.GasLimit, tracing.GasChangeTxInitialBalance)
+		}
+
+		tracer.OnTxStart(
+			evm.GetVMContext(),
+			ethtypes.NewTx(&ethtypes.LegacyTx{
+				To:    msg.To,
+				Data:  msg.Data,
+				Value: msg.Value,
+				Gas:   msg.GasLimit,
+			}),
+			msg.From,
+		)
+
+		defer func() {
+			debugFn()
+			tracer.OnTxEnd(&ethtypes.Receipt{GasUsed: gasUsed}, err)
+		}()
+
 		if cfg.DebugTrace {
 			amount := new(big.Int).Mul(msg.GasPrice, new(big.Int).SetUint64(msg.GasLimit))
-			stateDB.SubBalance(sender.Address(), amount)
+			stateDB.SubBalance(sender, uint256.MustFromBig(amount), tracing.BalanceChangeTransfer)
```

Comment:
```suggestion
			stateDB.SubBalance(sender, uint256.MustFromBig(amount), tracing.BalanceDecreaseGasBuy)
```

(we charge everything at the beginning)